pkg/liqo-controller-manager/namespace-controller/namespace_controller.go

```diff
@@ -0,0 +1,270 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace_controller
+
+import (
+	"context"
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+	liqo_controller_manager "github.com/liqotech/liqo/pkg/liqo-controller-manager"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/util/slice"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"strings"
+)
+
+type NamespaceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+const (
+	mappingLabel          = "mapping.liqo.io"
+	offloadingLabel       = "offloading.liqo.io"
+	offloadingPrefixLabel = "offloading.liqo.io/"
+	namespaceFinalizer    = "namespace.liqo.io/finalizer"
+)
+
+func (r *NamespaceReconciler) removeRemoteNamespace(localName string, nm *mapsv1alpha1.NamespaceMap) error {
+
+	if _, ok := nm.Status.NattingTable[localName]; ok {
+
+		delete(nm.Status.NattingTable, localName)
+		if err := r.Update(context.TODO(), nm); err != nil {
+			klog.Error(err, " --> Unable to update NamespaceMap")
+			return err
+		}
+		klog.Info(" Entries deleted correctly")
+
+	}
+
+	return nil
+}
+
+func (r *NamespaceReconciler) removeRemoteNamespaces(localName string, nms map[string]*mapsv1alpha1.NamespaceMap) error {
+
+	for _, nm := range nms {
+		if err := r.removeRemoteNamespace(localName, nm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *NamespaceReconciler) createRemoteNamespace(n *corev1.Namespace, remoteName string, nm *mapsv1alpha1.NamespaceMap) error {
+
+	if nm.Status.NattingTable == nil {
+		nm.Status.NattingTable = map[string]string{}
+	}
+
+	if oldValue, ok := nm.Status.NattingTable[n.GetName()]; ok {
+		if oldValue != remoteName {
+			n.GetLabels()[mappingLabel] = oldValue
+			if err := r.Update(context.TODO(), n); err != nil {
```

Comment:
Unfortunately, this is necessary if we want to manage the case in which a user try to change mapping.liqo.io label's value, we have to update again namespace with the old label's value..., an alternative way could be manage it in the event filter function `manageLabelPredicate`, but it doesn't seem a very refined solution...
Is not easy to factorize it because we have only 3 update case for namespace:
- first time, when finalizer is added.
- when finalizer is removed on deletion.
- and the case mentioned above.